internal/app/machined/pkg/controllers/network/dns_upstream.go

```diff
@@ -144,7 +130,43 @@ func (ctrl *DNSUpstreamController) run(ctx context.Context, r controller.Runtime
 	return nil
 }
 
-func (ctrl *DNSUpstreamController) cleanupUpstream(ctx context.Context, r controller.Runtime, touchedIDs map[resource.ID]struct{}, l *zap.Logger) {
+func existingConnections(ctx context.Context, r controller.Runtime) (func(*network.DNSUpstreamSpecSpec, string, *zap.Logger), error) {
+	upstream, err := safe.ReaderListAll[*network.DNSUpstream](ctx, r)
+	if err != nil {
+		return nil, err
+	}
+
+	existingConn := make(map[string]*network.DNSConn, upstream.Len())
+
+	for u := range upstream.All() {
+		existingConn[u.TypedSpec().Value.Conn.Addr()] = u.TypedSpec().Value.Conn
+	}
+
+	return func(spec *network.DNSUpstreamSpecSpec, remoteHost string, l *zap.Logger) {
+		remoteAddr := net.JoinHostPort(remoteHost, "53")
+		if spec.Conn != nil && spec.Conn.Addr() == remoteAddr {
+			l.Info("reusing existing upstream spec", zap.String("addr", remoteAddr))
```

Comment:
this feels like all should be `Debug`, nothing for the user to see here

Reply:
👍